Update 6 task

2022-tatsuki opened the pull request:
mainブランチを修正して日本以外の国でも返信できるようにしました。
また、時刻も表示できるようにしました。
ただ、一つのリクエストに1行しか返せないです。
もし、一つのリクエストに対して複数の返信をできる方法があれば教えてほしいです。
よろしくおねがします。

2022-tatsuki committed:
Update README.md

2022-tatsuki committed:
Update HelloWorldController.java

2022-tatsuki committed:
Update HelloWorldController.java

2022-tatsuki committed:
Delete helloworld/build directory

2022-tatsuki committed:
第6回課題修正版

不必要なファイルの削除
各国のあいさつについてはクエリパラメーターで受け取り、結果を返すようにしてみました。

YudaiNoda3576 commented on helloworld/src/main/java/com/example/helloworld/HelloWorldController.java:
```diff
@@ -15,24 +17,22 @@ public String hello() {
-    public Object us() {
-        String responseUs = "ようこそアメリカへ";
-        return responseUs;
+   @GetMapping("/greeting")
```
Restにおいてパスは複数形にするのが望ましいため複数形にしておきましょう！
また、その理由も共有してください！:pray:

YudaiNoda3576 commented on helloworld/src/main/java/com/example/helloworld/HelloWorldController.java:
```diff
@@ -15,24 +17,22 @@ public String hello() {
+    {
+        String greeting="Hello world";
+
+        if (Objects.equals(country,"japan")){
```
ここはcountryが渡されない場合があるのでガード節を使って早期returnしましょう〜

また、こういう分岐はswitch文の方がスッキリ書けそうですね！

今回くらいの条件分岐なら問題ないんですが練習も兼ねて。。。
Controllerにロジックを持ち込みたくないのと、返す言語を判定する処理はこのメソッドの振る舞いから若干逸脱しているように感じるので、Serviceクラスなどにこのロジックを移行してメソッドを切り離しておきましょう〜

こうすることでコードの可読性も上がるし、テストも書きやすくなります！

YudaiNoda3576 commented on helloworld/src/main/java/com/example/helloworld/HelloWorldController.java:
```diff
@@ -15,24 +17,22 @@ public String hello() {
+        }else if (Objects.equals(country,"france")){
+            greeting="Bonjour";
+        }else if (Objects.equals(country,"italy")){
+            greeting="Ciao";
```
nameが渡されているが、条件のどれにも該当するものがない場合の処理が抜けているように思うので、その処理を実装するようにお願いします:pray:

YudaiNoda3576 reviewed (changes requested):
コメントしています〜

2022-tatsuki committed:
2nd Update 6Task 

2回目の修正
クエリパラメータのメソッドを新たにServiceクラスに移し、ControllerクラスからServiceクラスのメソッドを呼びだした。

2022-tatsuki committed:
2nd update 6Task

新たにServiceクラスを作成し、そこにクエリパラメータで受け取るメソッドをControllerクラスから移した。

YudaiNoda3576 commented on helloworld/src/main/java/com/example/helloworld/HelloWorldController.java:
```diff
@@ -45,6 +26,11 @@ public Object japanTime() {
         return formatNowDate;
     }
 
+    @GetMapping("/greeting")
```
RequestParameterなどを受け取るのはControlelrの役割ですね！
なので、こちらに`@RequestParam(name="country",value ="country",defaultValue = "none",required = false)String country)`を付与するようにしましょう〜

YudaiNoda3576 commented on helloworld/src/main/java/com/example/helloworld/HelloWorldController.java:
```diff
@@ -45,6 +26,11 @@ public Object japanTime() {
+    @GetMapping("/greeting")
+    public String response(String country) {
+        Service responseService = new Service();
+        return responseService.worldGreeting(country);
```
そしてこちらのメソッドでcountryを受け取る必要はないので、仮引数のcountryは不要になるかと思います！
パラメータから受け取ったcountryをそのままServiceクラスのworldGreetingメソッドに渡してしまえばOKです！

YudaiNoda3576 reviewed (commented):
コメントしています！

YudaiNoda3576 commented on helloworld/src/main/java/com/example/helloworld/Service.java:
```diff
@@ -0,0 +1,24 @@
+
+public class Service {
+
+    public static String worldGreeting(@RequestParam(value = "country", defaultValue = "japan") String country) {
```
ここにrequestParamは不要ですね！

YudaiNoda3576 commented on helloworld/src/main/java/com/example/helloworld/Service.java:
```diff
@@ -0,0 +1,24 @@
+import org.springframework.web.bind.annotation.RequestParam;
+
+
+public class Service {
```
SpringBootでは@Serviceというアノテーションがあるので、そちらを使用しましょう！

YudaiNoda3576 reviewed (commented):
追加でコメントしています！

YudaiNoda3576 commented on helloworld/src/main/java/com/example/helloworld/HelloWorldController.java:
```diff
@@ -45,6 +26,11 @@ public Object japanTime() {
 
+    @GetMapping("/greeting")
+    public String response(String country) {
+        Service responseService = new Service();
```
ServiceクラスはConotroller全体で使用することになるので、クラスの全体に依存性を注入するようにしましょう！（方法は@Autowiredアノテーションを使うかコンストラクタで注入するかの二択ありますが、後者が望ましいです）

【余談】
理由としてはServiceクラスのインスタンスを生成する処理をそれぞれのメソッドでしてしまうと、同じクラスのオブジェクトが複数生成される場合があります。
そして、インスタンスの生成は非常にコストが高い処理です。
それが将来的にどんどん増えていくと考えると、今は問題ではなくとも将来のパフォーマンスの課題になりますよね？
そういったことを防ぐためにSpringでは依存性をクラス全体で注入することで一つのオブジェクトを生成するようにします（DI）
そしてSpringのDIはデフォルトでSingletonパターンと呼ばれるデザインパターンを採用しており、@Autowiredなどのアノテーションを付与することでそのオブジェクトが一つであるという状態を担保するように管理してくれます！

FYI
https://atmarkit.itmedia.co.jp/ait/articles/0504/29/news022.html

https://qiita.com/shoheiyokoyama/items/c16fd547a77773c0ccc1

2022-tatsuki committed:
Delete helloworld/src directory

2022-tatsuki committed:
3nd update

①@RequestParamをServiceクラスからControllerクラスへ移動させた。
②Serviceクラスで@serviceを使用した。
③ControllerクラスにSeviceクラスをDIするために＠Autowiredを使用した。
出力結果は前回と同じである。
③についてはできていない可能性がある。

YudaiNoda3576 commented on helloworld/src/main/java/com/example/helloworld/Service.java:
```diff
@@ -0,0 +1,24 @@
+
+import org.springframework.stereotype.Component;
+
+@Component
```
@Serviceは使用できませんでしたか？？

2022-tatsuki replied:
しました。
その次の行に入れていたのですが・・・・
調べると似たような役割？と出てきたので片方だけでいいのでしょうか？

YudaiNoda3576 replied:
@Serviceと@Componentの役割について調べてみる必要がありそうですね！

2022-tatsuki replied:
https://qiita.com/KevinFQ/items/abc7369cb07eb4b9ae29
このサイトを見ると@serviceと＠componentの動きは同じで使い分けはそのクラスがどんな役割があるのかによって使い分ける。
今回で言うと、Serviceクラスには＠serviceをつけ、HelloWorldControllerには＠RestControllerをつけるのがいいと思うのですが、あってますでしょうか？

YudaiNoda3576 replied:
その認識であっています〜🙌

2022-tatsuki replied:
ありがとうございます！

YudaiNoda3576 commented on helloworld/src/main/java/com/example/helloworld/HelloWorldController.java:
```diff
@@ -1,20 +1,40 @@
+
 @RestController
 public class HelloWorldController {
+    @Autowired
```
ここはこの辺りの記事を参考にして見比べてみてください！
https://www.shookuro.com/entry/2016/08/09/175801

また、コンストラクタインジェクションが推奨されているので、実装で使用しないにせよ
推奨されている理由は調べて共有して欲しいです~!

2022-tatsuki replied:
https://www.shookuro.com/entry/2016/08/09/175801
https://pppurple.hatenablog.com/entry/2016/12/29/233141
この2つのサイトを調べました。

コンストラクタインジェクションは@Autowairedに比べると
△記述するコードが多くなる
○その分単体のテストがしやすい。
○クラスに多くの依存関係があることを知らせてくれて、依存関係が多い場合、そのクラスが多くの責任を持ちすぎてる可能性があることを知らせてくれる。
○finalを使うことができる。
などのことが書いてありました。この認識で合っているのでしょうか？

YudaiNoda3576 replied:
はい！
大丈夫そうです！
今はピンとこないかもしれないですが、このことは頭の片隅においておいてください！

2022-tatsuki replied:
わかりました！

YudaiNoda3576 reviewed (commented):
コメントしています〜！

2022-tatsuki committed:
4nd up date

2022-tatsuki committed:
4nd update

YudaiNoda3576 commented on helloworld/src/main/java/com/example/helloworld/Service.java:
```diff
@@ -1,13 +1,10 @@
-import org.springframework.stereotype.Component;
-
-@Component
 @org.springframework.stereotype.Service
```
ここは@Serviceとするだけで大丈夫です！

2022-tatsuki replied:
わかりました！

YudaiNoda3576 reviewed (commented):
コメントしています！

YudaiNoda3576 commented on helloworld/src/main/java/com/example/helloworld/HelloWorldController.java:
```diff
@@ -29,11 +35,18 @@ public Object japanTime() {
 
     @GetMapping("/greeting")
-    public String response(@RequestParam(value = "country", defaultValue = "japan") String country) {
+    public String getService(@RequestParam(value = "country", defaultValue = "japan") String country) {
```
このメソッド名は`getService`が適切でしょうか？
メソッドはプログラムの振る舞いなので、ここがやっていることを一言で表すような命名が良いかと思います！
returnしている結果が何を返しているのか？に着目すると適切な命名になるかと思います！

2022-tatsuki replied:
greetingMethod　にしました。
どうでしょうか？

YudaiNoda3576 replied:
methodであることは自明なので、methodと記載する必要はないかなと思います！
なのでシンプルにgreetとかで良いかなと思います！
メソッド名やクラス名、コメントなどはそれ自体がシステムの仕様書みたいな存在でもあるので、それを見た時に開発者が有益な情報を得られるように工夫するのが良いです！
そういったことを学ぶ書籍としておすすめなのが[リーダブルコード](https://www.amazon.co.jp/%E3%83%AA%E3%83%BC%E3%83%80%E3%83%96%E3%83%AB%E3%82%B3%E3%83%BC%E3%83%89-%E2%80%95%E3%82%88%E3%82%8A%E8%89%AF%E3%81%84%E3%82%B3%E3%83%BC%E3%83%89%E3%82%92%E6%9B%B8%E3%81%8F%E3%81%9F%E3%82%81%E3%81%AE%E3%82%B7%E3%83%B3%E3%83%97%E3%83%AB%E3%81%A7%E5%AE%9F%E8%B7%B5%E7%9A%84%E3%81%AA%E3%83%86%E3%82%AF%E3%83%8B%E3%83%83%E3%82%AF-Theory-practice-Boswell/dp/4873115655/ref=sr_1_1?adgrpid=52747835709&gclid=CjwKCAiA8OmdBhAgEiwAShr406DGSWTsSoAc6FM9ATx9FqPsiuthZFjpNcqXp3ZEl-cnGg_osyb7_xoCnTwQAvD_BwE&hvadid=553962830540&hvdev=c&hvlocphy=1009497&hvnetw=g&hvqmt=e&hvrand=6086049033667657001&hvtargid=kwd-334307148361&hydadcr=16038_13486739&jp-ad-ap=0&keywords=%E3%83%AA%E3%83%BC%E3%83%80%E3%83%96%E3%83%AB%E3%82%B3%E3%83%BC%E3%83%89&qid=1673245124&sr=8-1)なのでこちらよければ読んでみてください🙌

2022-tatsuki replied:
わかりました！
ありがとうございます。

YudaiNoda3576 reviewed (commented):
コメントしています！

2022-tatsuki committed:
Update Service.java

2022-tatsuki committed:
4nd update

細かい部分の修正とコンストラクタインジェクションをしてみました。

2022-tatsuki committed:
Update HelloWorldController.java
